[bug] `AnalysisException` when materializing a DataFrame containing `F.col('array_col.struct_field')` in its plan

### Environment

- Sail version: 0.6.3
- pyspark-client: 4.1.1
- Platform: Windows 11

### Description

Accessing `df.columns` raises `AnalysisException` when the DataFrame's logical plan contains a **dotted column reference** of the form `F.col('array_col.struct_field')` where `array_col` is `ArrayType(StructType(...))`.

This is valid, working Spark behavior in classic PySpark: the dotted notation projects the named field out of each struct element, yielding an `ArrayType(primitive)`.

### Minimal reproducible example

```python
from pyspark.sql import SparkSession
from pyspark.sql import functions as F
from pyspark.sql import types as T

spark = SparkSession.builder.remote("sc://localhost").getOrCreate()

schema = T.StructType(
    [
        T.StructField("id", T.StringType()),
        T.StructField("outer_value", T.IntegerType()),
        T.StructField(
            "array_of_structs",
            T.ArrayType(T.StructType([T.StructField("inner_value", T.IntegerType())])),
        ),
    ]
)
df = spark.createDataFrame(
    schema=schema,
    data=[
        ("row1", 42, [{"inner_value": 42}]),
        ("row2", 99, [{"inner_value": 1}]),
    ],
)
df2 = df.withColumn(
    "contains",
    F.array_contains(F.col("array_of_structs.inner_value"), F.col("outer_value")),  # works fine in JVM PySpark
)

print(df2.columns)  # ❌ raises AnalysisException in LakeSail
```

### Expected behavior

`df2.columns` returns `["id", "outer_value", "array_of_structs", "contains"]`.

`array_of_structs` is present in the plan; `F.col("array_of_structs.inner_value")` is a supported Spark shorthand for projecting a field out of an array of structs, producing `ArrayType(IntegerType())`, which is a valid first argument to `array_contains`.

### Actual behavior
`pyspark.errors.exceptions.connect.AnalysisException: attribute ObjectName([Identifier("array_of_structs"), Identifier("inner_value")]) is missing from the schema: cannot resolve attribute`

<details>

<summary>stack trace</summary>

```python
.venv\lib\site-packages\pyspark\sql\connect\dataframe.py:306: in columns
    return [field.name for field in self._schema.fields]
.venv\lib\site-packages\pyspark\sql\connect\dataframe.py:1884: in _schema
    self._cached_schema = self._session.client.schema(query)
.venv\lib\site-packages\pyspark\sql\connect\client\core.py:1159: in schema
    schema = self._analyze(method="schema", plan=plan).schema
.venv\lib\site-packages\pyspark\sql\connect\client\core.py:1430: in _analyze
    self._handle_error(error)
.venv\lib\site-packages\pyspark\sql\connect\client\core.py:1982: in _handle_error
    self._handle_rpc_error(error)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pyspark.sql.connect.client.core.SparkConnectClient object at 0x00000241B5960BE0>
rpc_error = <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.INTERNAL
        details = "attribute ObjectName([Identi..._of_structs\"), Identifier(\"inner_value\")]) is missing from the schema: cannot resolve attribute", grpc_status:13}"
>

    def _handle_rpc_error(self, rpc_error: grpc.RpcError) -> NoReturn:
        """
        Error handling helper for dealing with GRPC Errors. On the server side, certain
        exceptions are enriched with additional RPC Status information. These are
        unpacked in this function and put into the exception.
    
        To avoid overloading the user with GRPC errors, this message explicitly
        swallows the error context from the call. This GRPC Error is logged however,
        and can be enabled.
    
        Parameters
        ----------
        rpc_error : grpc.RpcError
           RPC Error containing the details of the exception.
    
        Returns
        -------
        Throws the appropriate internal Python exception.
        """
        logger.exception("GRPC Error received")
        # We have to cast the value here because, a RpcError is a Call as well.
        # https://grpc.github.io/grpc/python/grpc.html#grpc.UnaryUnaryMultiCallable.__call__
        error: grpc.Call = cast(grpc.Call, rpc_error)
        status_code: grpc.StatusCode = error.code()
        status: Optional[Status] = rpc_status.from_call(error)
        if status:
            for d in status.details:
                if d.Is(error_details_pb2.ErrorInfo.DESCRIPTOR):
                    info = error_details_pb2.ErrorInfo()
                    d.Unpack(info)
                    logger.debug(f"Received ErrorInfo: {info}")
    
                    if info.metadata.get("errorClass") == "INVALID_HANDLE.SESSION_CHANGED":
                        self._closed = True
    
                    if info.metadata.get("errorClass") == "CONNECT_INVALID_PLAN.CANNOT_PARSE":
                        # Disable plan compression if the server fails to interpret the plan.
                        logger.info(
                            "Disabling plan compression for the session due to "
                            "CONNECT_INVALID_PLAN.CANNOT_PARSE error."
                        )
                        self._plan_compression_threshold, self._plan_compression_algorithm = (
                            -1,
                            "NONE",
                        )
    
>                   raise convert_exception(
                        info,
                        status.message,
                        self._fetch_enriched_error(info),
                        self._display_server_stack_trace(),
                        status_code,
                    ) from None
E                   pyspark.errors.exceptions.connect.AnalysisException: attribute ObjectName([Identifier("array_of_structs"), Identifier("inner_value")]) is missing from the schema: cannot resolve attribute
```

</details>

--- a/crates/sail-execution/src/codec.rs
+++ b/crates/sail-execution/src/codec.rs
@@ -124,6 +124,7 @@ use sail_function::scalar::array::spark_array_compact::SparkArrayCompact;
 use sail_function::scalar::array::spark_array_item_with_position::ArrayItemWithPosition;
 use sail_function::scalar::array::spark_array_min_max::{ArrayMax, ArrayMin};
 use sail_function::scalar::array::spark_sequence::SparkSequence;
+use sail_function::scalar::array_struct_field::ArrayStructField;
 use sail_function::scalar::collection::spark_concat::SparkConcat;
 use sail_function::scalar::collection::spark_reverse::SparkReverse;
 use sail_function::scalar::csv::spark_from_csv::SparkFromCSV;
@@ -2231,6 +2232,7 @@ impl PhysicalExtensionCodec for RemoteExecutionCodec {
             "array_item_with_position" => {
                 Ok(Arc::new(ScalarUDF::from(ArrayItemWithPosition::new())))
             }
+            "array_struct_field" => Ok(Arc::new(ScalarUDF::from(ArrayStructField::new()))),
             "array_min" => Ok(Arc::new(ScalarUDF::from(ArrayMin::new()))),
             "array_max" => Ok(Arc::new(ScalarUDF::from(ArrayMax::new()))),
             "array_intersect" | "list_intersect" => {
@@ -2410,6 +2412,7 @@ impl PhysicalExtensionCodec for RemoteExecutionCodec {
         // TODO: Implement custom registry to avoid codec for built-in functions
         let node_inner = node.inner().as_any();
         let udf_kind: UdfKind = if node_inner.is::<ArrayItemWithPosition>()
+            || node_inner.is::<ArrayStructField>()
             || node_inner.is::<ArrayMax>()
             || node_inner.is::<ArrayMin>()
             || node_inner.is::<ArrayIntersect>()

--- a/crates/sail-function/src/scalar/array_struct_field.rs
+++ b/crates/sail-function/src/scalar/array_struct_field.rs
@@ -7,7 +7,9 @@ use datafusion::arrow::array::{
 use datafusion::arrow::buffer::NullBuffer;
 use datafusion::arrow::datatypes::{DataType, Field, FieldRef};
 use datafusion_common::cast::{as_large_list_array, as_list_array};
-use datafusion_common::{exec_err, internal_err, plan_datafusion_err, Result, ScalarValue};
+use datafusion_common::{
+    exec_datafusion_err, exec_err, internal_err, plan_datafusion_err, Result, ScalarValue,
+};
 use datafusion_expr::{
     ColumnarValue, ReturnFieldArgs, ScalarFunctionArgs, ScalarUDFImpl, Signature, Volatility,
 };
@@ -48,17 +50,6 @@ impl ArrayStructField {
         }
     }
 
-    fn child_field<'a>(fields: &'a [FieldRef], name: &str) -> Result<&'a FieldRef> {
-        let matches = fields
-            .iter()
-            .filter(|field| field.name().eq_ignore_ascii_case(name))
-            .collect::<Vec<_>>();
-        let [field] = matches.as_slice() else {
-            return exec_err!("missing or ambiguous field: {name}");
-        };
-        Ok(field)
-    }
-
     fn return_type_for_list(&self, data_type: &DataType, field_name: &str) -> Result<DataType> {
         let list_field = Self::list_field(data_type)?;
         let DataType::Struct(fields) = list_field.data_type() else {
@@ -67,7 +58,9 @@ impl ArrayStructField {
                 list_field.data_type()
             );
         };
-        let child = Self::child_field(fields, field_name)?;
+        let (_, child) = fields
+            .find(field_name)
+            .ok_or_else(|| exec_datafusion_err!("missing field: {field_name}"))?;
         let nullable = list_field.is_nullable() || child.is_nullable();
         let field = Arc::new(Field::new_list_field(child.data_type().clone(), nullable));
         match data_type {
@@ -78,36 +71,34 @@ impl ArrayStructField {
         }
     }
 
-    fn struct_values(values: &ArrayRef, field_name: &str) -> Result<ArrayRef> {
+    fn struct_values(values: &ArrayRef, field_name: &str) -> Result<(ArrayRef, bool)> {
         let Some(struct_array) = values.as_any().downcast_ref::<StructArray>() else {
             return exec_err!(
                 "array_struct_field requires list values to be structs, got {}",
                 values.data_type()
             );
         };
-        let matches = struct_array
+        let (index, field) = struct_array
             .fields()
-            .iter()
-            .enumerate()
-            .filter(|(_, field)| field.name().eq_ignore_ascii_case(field_name))
-            .map(|(index, _)| index)
-            .collect::<Vec<_>>();
-        let [index] = matches.as_slice() else {
-            return exec_err!("missing or ambiguous field: {field_name}");
-        };
-        let values = struct_array.column(*index);
+            .find(field_name)
+            .ok_or_else(|| exec_datafusion_err!("missing field: {field_name}"))?;
+        let values = struct_array.column(index);
         let nulls = NullBuffer::union(struct_array.nulls(), values.nulls());
         let data = values.to_data().into_builder().nulls(nulls).build()?;
-        Ok(make_array(data))
+        Ok((make_array(data), field.is_nullable()))
     }
 
     fn project_list<O>(array: &GenericListArray<O>, field_name: &str) -> Result<ArrayRef>
     where
         O: OffsetSizeTrait,
     {
-        let values = Self::struct_values(array.values(), field_name)?;
+        let list_field = Self::list_field(array.data_type())?;
+        let (values, child_nullable) = Self::struct_values(array.values(), field_name)?;
         Ok(Arc::new(GenericListArray::<O>::try_new(
-            Arc::new(Field::new_list_field(values.data_type().clone(), true)),
+            Arc::new(Field::new_list_field(
+                values.data_type().clone(),
+                list_field.is_nullable() || child_nullable,
+            )),
             array.offsets().clone(),
             values,
             array.nulls().cloned(),
@@ -119,9 +110,13 @@ impl ArrayStructField {
         size: i32,
         field_name: &str,
     ) -> Result<ArrayRef> {
-        let values = Self::struct_values(array.values(), field_name)?;
+        let list_field = Self::list_field(array.data_type())?;
+        let (values, child_nullable) = Self::struct_values(array.values(), field_name)?;
         Ok(Arc::new(FixedSizeListArray::try_new(
-            Arc::new(Field::new_list_field(values.data_type().clone(), true)),
+            Arc::new(Field::new_list_field(
+                values.data_type().clone(),
+                list_field.is_nullable() || child_nullable,
+            )),
             size,
             values,
             array.nulls().cloned(),

--- a/crates/sail-function/src/scalar/array_struct_field.rs
+++ b/crates/sail-function/src/scalar/array_struct_field.rs
@@ -1,0 +1,202 @@
+use std::any::Any;
+use std::sync::Arc;
+
+use datafusion::arrow::array::{
+    make_array, Array, ArrayRef, FixedSizeListArray, GenericListArray, OffsetSizeTrait, StructArray,
+};
+use datafusion::arrow::buffer::NullBuffer;
+use datafusion::arrow::datatypes::{DataType, Field, FieldRef};
+use datafusion_common::cast::{as_large_list_array, as_list_array};
+use datafusion_common::{exec_err, internal_err, plan_datafusion_err, Result, ScalarValue};
+use datafusion_expr::{
+    ColumnarValue, ReturnFieldArgs, ScalarFunctionArgs, ScalarUDFImpl, Signature, Volatility,
+};
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub struct ArrayStructField {
+    signature: Signature,
+}
+
+impl Default for ArrayStructField {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ArrayStructField {
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::any(2, Volatility::Immutable),
+        }
+    }
+
+    fn field_name<'a>(&self, value: Option<&'a ScalarValue>) -> Result<&'a str> {
+        value
+            .and_then(|x| x.try_as_str().flatten())
+            .filter(|x| !x.is_empty())
+            .ok_or_else(|| {
+                plan_datafusion_err!("{} requires a non-empty literal field name", self.name())
+            })
+    }
+
+    fn list_field(data_type: &DataType) -> Result<&FieldRef> {
+        match data_type {
+            DataType::List(field)
+            | DataType::LargeList(field)
+            | DataType::FixedSizeList(field, _) => Ok(field),
+            _ => exec_err!("array_struct_field requires a list, got {data_type}"),
+        }
+    }
+
+    fn child_field<'a>(fields: &'a [FieldRef], name: &str) -> Result<&'a FieldRef> {
+        let matches = fields
+            .iter()
+            .filter(|field| field.name().eq_ignore_ascii_case(name))
+            .collect::<Vec<_>>();
+        let [field] = matches.as_slice() else {
+            return exec_err!("missing or ambiguous field: {name}");
+        };
+        Ok(field)
+    }
+
+    fn return_type_for_list(&self, data_type: &DataType, field_name: &str) -> Result<DataType> {
+        let list_field = Self::list_field(data_type)?;
+        let DataType::Struct(fields) = list_field.data_type() else {
+            return exec_err!(
+                "array_struct_field requires a list of structs, got {}",
+                list_field.data_type()
+            );
+        };
+        let child = Self::child_field(fields, field_name)?;
+        let nullable = list_field.is_nullable() || child.is_nullable();
+        let field = Arc::new(Field::new_list_field(child.data_type().clone(), nullable));
+        match data_type {
+            DataType::List(_) => Ok(DataType::List(field)),
+            DataType::LargeList(_) => Ok(DataType::LargeList(field)),
+            DataType::FixedSizeList(_, size) => Ok(DataType::FixedSizeList(field, *size)),
+            _ => unreachable!("list_field accepted only list data types"),
+        }
+    }
+
+    fn struct_values(values: &ArrayRef, field_name: &str) -> Result<ArrayRef> {
+        let Some(struct_array) = values.as_any().downcast_ref::<StructArray>() else {
+            return exec_err!(
+                "array_struct_field requires list values to be structs, got {}",
+                values.data_type()
+            );
+        };
+        let matches = struct_array
+            .fields()
+            .iter()
+            .enumerate()
+            .filter(|(_, field)| field.name().eq_ignore_ascii_case(field_name))
+            .map(|(index, _)| index)
+            .collect::<Vec<_>>();
+        let [index] = matches.as_slice() else {
+            return exec_err!("missing or ambiguous field: {field_name}");
+        };
+        let values = struct_array.column(*index);
+        let nulls = NullBuffer::union(struct_array.nulls(), values.nulls());
+        let data = values.to_data().into_builder().nulls(nulls).build()?;
+        Ok(make_array(data))
+    }
+
+    fn project_list<O>(array: &GenericListArray<O>, field_name: &str) -> Result<ArrayRef>
+    where
+        O: OffsetSizeTrait,
+    {
+        let values = Self::struct_values(array.values(), field_name)?;
+        Ok(Arc::new(GenericListArray::<O>::try_new(
+            Arc::new(Field::new_list_field(values.data_type().clone(), true)),
+            array.offsets().clone(),
+            values,
+            array.nulls().cloned(),
+        )?))
+    }
+
+    fn project_fixed_size_list(
+        array: &FixedSizeListArray,
+        size: i32,
+        field_name: &str,
+    ) -> Result<ArrayRef> {
+        let values = Self::struct_values(array.values(), field_name)?;
+        Ok(Arc::new(FixedSizeListArray::try_new(
+            Arc::new(Field::new_list_field(values.data_type().clone(), true)),
+            size,
+            values,
+            array.nulls().cloned(),
+        )?))
+    }
+}
+
+impl ScalarUDFImpl for ArrayStructField {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "array_struct_field"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, _arg_types: &[DataType]) -> Result<DataType> {
+        internal_err!("return_field_from_args should be called instead")
+    }
+
+    fn return_field_from_args(&self, args: ReturnFieldArgs) -> Result<FieldRef> {
+        let [array_field, _] = args.arg_fields else {
+            return exec_err!(
+                "{} requires 2 arguments, got {}",
+                self.name(),
+                args.arg_fields.len()
+            );
+        };
+        let field_name = self.field_name(args.scalar_arguments.get(1).copied().flatten())?;
+        let data_type = self.return_type_for_list(array_field.data_type(), field_name)?;
+        Ok(Arc::new(Field::new(self.name(), data_type, true)))
+    }
+
+    fn invoke_with_args(&self, args: ScalarFunctionArgs) -> Result<ColumnarValue> {
+        let ScalarFunctionArgs {
+            args, arg_fields, ..
+        } = args;
+        let field_name = arg_fields
+            .get(1)
+            .and_then(|_| args.get(1))
+            .and_then(|arg| match arg {
+                ColumnarValue::Scalar(value) => value.try_as_str().flatten(),
+                ColumnarValue::Array(_) => None,
+            })
+            .filter(|x| !x.is_empty())
+            .ok_or_else(|| {
+                plan_datafusion_err!("{} requires a non-empty literal field name", self.name())
+            })?;
+        let arrays = ColumnarValue::values_to_arrays(&args[..1])?;
+        let [array] = arrays.as_slice() else {
+            return exec_err!("{} requires an array argument", self.name());
+        };
+        let out = match array.data_type() {
+            DataType::List(_) => Self::project_list(as_list_array(array)?, field_name)?,
+            DataType::LargeList(_) => Self::project_list(as_large_list_array(array)?, field_name)?,
+            DataType::FixedSizeList(_, size) => {
+                let Some(array) = array.as_any().downcast_ref::<FixedSizeListArray>() else {
+                    return exec_err!(
+                        "array_struct_field expected FixedSizeListArray, got {}",
+                        array.data_type()
+                    );
+                };
+                Self::project_fixed_size_list(array, *size, field_name)?
+            }
+            _ => {
+                return exec_err!(
+                    "array_struct_field requires a list, got {}",
+                    array.data_type()
+                )
+            }
+        };
+        Ok(ColumnarValue::Array(out))
+    }
+}

--- a/crates/sail-function/src/scalar/mod.rs
+++ b/crates/sail-function/src/scalar/mod.rs
@@ -1,4 +1,5 @@
 pub mod array;
+pub mod array_struct_field;
 pub mod collection;
 pub mod csv;
 pub mod datetime;

--- a/crates/sail-plan/src/resolver/expression/attribute.rs
+++ b/crates/sail-plan/src/resolver/expression/attribute.rs
@@ -1,9 +1,12 @@
-use arrow::datatypes::DataType;
+use std::sync::Arc;
+
+use arrow::datatypes::{DataType, Field};
 use datafusion_common::{Column, DFSchemaRef, TableReference};
 use datafusion_expr::expr::ScalarFunction;
-use datafusion_expr::{col, expr, lit};
+use datafusion_expr::{col, expr, lit, ScalarUDF};
 use datafusion_functions::core::get_field;
 use sail_common::spec;
+use sail_function::scalar::array_struct_field::ArrayStructField;
 
 use crate::error::{PlanError, PlanResult};
 use crate::resolver::expression::NamedExpr;
@@ -234,6 +237,33 @@ impl PlanResolver<'_> {
                             expr::Expr::ScalarFunction(ScalarFunction::new_udf(get_field(), args));
                         Self::resolve_potentially_nested_field(expr, field.data_type(), remaining)
                     }),
+                DataType::List(field)
+                | DataType::LargeList(field)
+                | DataType::FixedSizeList(field, _) => {
+                    let DataType::Struct(fields) = field.data_type() else {
+                        return None;
+                    };
+                    fields
+                        .iter()
+                        .find(|x| x.name().eq_ignore_ascii_case(name.as_ref()))
+                        .and_then(|child| {
+                            let expr = ScalarUDF::from(ArrayStructField::new())
+                                .call(vec![expr, lit(child.name().to_string())]);
+                            let item = Arc::new(Field::new_list_field(
+                                child.data_type().clone(),
+                                field.is_nullable() || child.is_nullable(),
+                            ));
+                            let data_type = match data_type {
+                                DataType::List(_) => DataType::List(item),
+                                DataType::LargeList(_) => DataType::LargeList(item),
+                                DataType::FixedSizeList(_, size) => {
+                                    DataType::FixedSizeList(item, *size)
+                                }
+                                _ => unreachable!("list data type matched above"),
+                            };
+                            Self::resolve_potentially_nested_field(expr, &data_type, remaining)
+                        })
+                }
                 _ => None,
             },
         }

--- a/python/pysail/tests/spark/test_column.py
+++ b/python/pysail/tests/spark/test_column.py
@@ -101,3 +101,30 @@ def test_try_cast_invalid_date(spark):
     ).collect()
 
     assert result == [Row(id=None, date_col=None, ts_col=None, ts_ntz_col=None)]
+
+
+def test_array_struct_field(spark):
+    df = spark.createDataFrame(
+        data=[
+            ("0", [{"b": 42, "c": {"d": 100.0}}]),
+            ("1", [{"b": 1, "c": None}, {"b": -1}]),
+            ("2", [None, {"b": 3, "c": {"d": None}}, {"b": None}]),
+            ("3", None),
+        ],
+        schema="id: string, a: array<struct<b: int, c: struct<d: double>>>",
+    )
+    actual = df.select("id", F.col("a.b")).collect()
+    assert sorted(actual, key=lambda row: row.id) == [
+        Row(id="0", b=[42]),
+        Row(id="1", b=[1, -1]),
+        Row(id="2", b=[None, 3, None]),
+        Row(id="3", b=None),
+    ]
+
+    actual = df.select("id", F.col("a.c.d")).collect()
+    assert sorted(actual, key=lambda row: row.id) == [
+        Row(id="0", d=[100.0]),
+        Row(id="1", d=[None, None]),
+        Row(id="2", d=[None, None, None]),
+        Row(id="3", d=None),
+    ]

